### PR TITLE
Proof-of-concept implementation of faster GeometrySystemGenerator

### PR DESCRIPTION
**Pull request by jchodera**
@pgrinaway had noted that the energy computation of the special `CustomForce`-encoded `System` object generated by `GeometrySystemGenerator` was a rate-limiting step that would likely make particle filtering using our current approach of drawing from `p(torsion | bond, angle)` impractical, so I attempted a different scheme to try to speed this up.

Instead of using `CustomForce` forces, this proof-of-concept implementation uses a copy of all the standard forces, and uses `updateParticleInContext` to update the parameters in the active `Context` when the particle index is incremented. This implementation is still lacking the addition of the few ring-closure torsions and angles, but these can be easily added. The implementation can also likely be sped up further by precomputing the growth indices of all bonds, angles, torsions, and nonbonded particles and exceptions.

I refactored `GeometrySystemGenerator` in order to move the function that updates the growth index for the active context in this class. That allows us to switch between the standard `GeometrySystemGenerator` (`CustomForce` version) and the new `GeometrySystemGeneratorFast` class. We can probably change this to a flag or something if we want to make it even simpler to compare.

Preliminary tests indicate the new implementation is much faster. For `ValenceSmallMoleculeLibraryTestSystem`, the geometry step for forward proposals goes from 2.4 s to 0.7 s. Commenting out the energy calculation completely in `_torsion_log_pmf` takes 0.3 s, suggesting the energy calculation is now on par with all the other overhead for this simple valence system.

For the `KinaseInhibitorsTestSystem` in vacuum, forward geometry proposals went from 10.8 s forward / 13.7 s reverse to 3.6 s / 4.4 s.

I couldn't try naphthalene or other recent test systems because I started from the `master` branch and all of those other systems are in their own branches.

### CODE REVIEW

**Comment by jchodera** [comment]
Also, tests would need updating since #270 hasn't been merged.

**Comment by jchodera** [comment]
To make this faster and more scalable, we could:
* pre-compile lists of all bonds, angles, torsions, and nonbonded particles/exceptions involved in each growth index increment, and only update those (instead of recomputing the growth index for every term each time)
* omit bonds, angles, and torsions that are not involved in particle growth from the `_reference_system` 
* reduce the cutoff or change the `nonbondedMethod` when `sterics` are in use

**Comment by jchodera** [comment]
The travis failure is due to the same issue as https://github.com/choderalab/alchemy/pull/53/files

File: perses/rjmc/geometry.py [hunk 1]
```diff
@@ -2149,6 +2162,115 @@ def _calculate_growth_idx(self, particle_indices, growth_indices):
...
+    def set_growth_parameter_index(self, growth_index, context=None):
+        """
+        Set the growth parameter index
+        """
+        for (growth_force, reference_force) in zip(self._growth_system.getForces(), self._reference_system.getForces()):
+            force_name = growth_force.__class__.__name__
+            if (force_name == 'HarmonicBondForce'):
+                for bond in range(reference_force.getNumBonds()):
+                    parameters = reference_force.getBondParameters(bond)
+                    this_growth_index = self._calculate_growth_idx(parameters[:2], self._growth_indices)
```
Comment (jchodera) [inline perses/rjmc/geometry.py]: You can precalculate these growth indices for each of the bonds, angles, torsions, particles, and exceptions in order to further speed this up.

If you're always incrementing the growth index by 1, you can also save time by only updating the parameters that are changing.

**Comment by jchodera** [comment]
Looks like you have to rebase or merge from `master` to resolve some conflicts.

**Comment by pgrinaway** [comment]
Yep, doing that now.

**Comment by pgrinaway** [comment]
Doing that again...

**Comment by pgrinaway** [comment]
I messed up when merging stuff I think. Will fix now.

File: perses/rjmc/geometry.py [hunk 2]
```diff
@@ -2142,9 +2142,9 @@ def _determine_extra_angles(self, angle_force, reference_topology, growth_indice
                     growth_idx = self._calculate_growth_idx(atom_indices, growth_indices)
                     #If this is a CustomAngleForce, we need to pass the parameters as a list, and it will have the growth_idx parameter.
                     #If it's a regular HarmonicAngleForce, there is no growth_index and the parameters are passed separately.
-                    if type(angle_force) == openmm.CustomAngleForce:
+                    if isinstance(angle_force, openmm.CustomAngleForce):
                         angle_force.addAngle(atom_indices[0], atom_indices[1], atom_indices[2], [angle_radians, angle_force_constant, growth_idx])
-                    elif type(angle_force) == openmm.HarmonicAngleForce:
+                    elif isinstance(angle_force, openmm.HarmonicAngleForce)
```
Comment (jchodera) [inline perses/rjmc/geometry.py]: Forgot colon at the end of the line!

**Comment by pgrinaway** [comment]
The API change was not fully accommodated (instead of `context.setParameter` we call a method of the `GeometrySystemGenerator` to handle the case where there is no `growth_index`)

**Comment by pgrinaway** [comment]
Ok, I think this PR is good to merge.

**Pull request merged**